File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        // Step 1: Count frequencies using HashMap
        HashMap<Integer, Integer> freqMap = new HashMap<>();
        for (int num : nums) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }

        // Step 2: Group numbers by frequency using TreeMap (descending order)
        TreeMap<Integer, List<Integer>> bucket = new TreeMap<>(Collections.reverseOrder());
        for (int num : freqMap.keySet()) {
            int freq = freqMap.get(num);
            bucket.putIfAbsent(freq, new ArrayList<>());
            bucket.get(freq).add(num);
        }

        // Step 3: Collect top K frequent elements
        List<Integer> result = new ArrayList<>();
        for (int freq : bucket.keySet()) {
            for (int num : bucket.get(freq)) {
                result.add(num);
                k--;
                if (k == 0) break; // break inner
            }
            if (k == 0) break; // break outer as well!
        }

        // Step 4: Convert to int[]
        int[] ans = new int[result.size()];
        for (int i = 0; i < result.size(); i++) {
            ans[i] = result.get(i);
        }

        return ans;
    }
}

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        
        freqMap = Counter(nums)
        buckets = [[] for _ in range(len(nums)+1)]

        for num,freq in freqMap.items():
            buckets[freq].append(num)

        # try to retrive k elements from it by traversing in reverse order
        res = []
        for i in range(len(buckets)-1,0,-1):
            for n in buckets[i]:
                res.append(n)
                if len(res) == k:
                    return res




        